helm-ag.el

```diff
@@ -618,40 +618,44 @@ Default behaviour shows finish and result in mode-line."
   (interactive)
   (goto-char (point-min))
   (let ((read-only-files 0)
-        (saved-buffers nil)
+        (files-to-lines (make-hash-table))
         (regexp (helm-ag--match-line-regexp))
-        (default-directory helm-ag--default-directory)
-        (line-deletes (make-hash-table :test #'equal))
-        (kept-buffers (buffer-list))
-        open-buffers)
+        (line-deletes (make-hash-table :test #'equal)))
+    ;; Group changes by file
     (while (re-search-forward regexp nil t)
       (let* ((file (or (match-string-no-properties 1) helm-ag--search-this-file-p))
              (line (string-to-number (match-string-no-properties 2)))
              (body (match-string-no-properties 3))
-             (ovs (overlays-at (line-beginning-position))))
-        (with-current-buffer (find-file-noselect file)
-          (cl-pushnew (current-buffer) open-buffers)
-          (if buffer-read-only
-              (cl-incf read-only-files)
-            (goto-char (point-min))
-            (let ((deleted-lines (gethash file line-deletes 0))
-                  (deleted (and ovs (overlay-get (car ovs) 'helm-ag-deleted))))
-              (forward-line (- line 1 deleted-lines))
-              (delete-region (line-beginning-position) (line-end-position))
-              (if (not deleted)
-                  (insert body)
-                (let ((beg (point)))
-                  (forward-line 1)
-                  (delete-region beg (point))
-                  (puthash file (1+ deleted-lines) line-deletes)))
-              (cl-pushnew (current-buffer) saved-buffers))))))
-    (when helm-ag-edit-save
-      (dolist (buf saved-buffers)
-        (with-current-buffer buf
-          (save-buffer))))
-    (dolist (buf open-buffers)
-      (unless (memq buf kept-buffers)
-        (kill-buffer buf)))
+             (ovs (overlays-at (line-beginning-position)))
+             (lines-list (gethash file files-to-lines)))
```

Comment:
Just added a variable to the let* statement to store a list of changed lines for each file.